Fix/settings page flow

TuxedoFish opened the pull request:
## What does this PR do?
Fixes #2110 so that you can log in by clicking on the settings page. #2117 Has been resolved in an older commit lost where it got merged in.

## ✅ Pull Request checklist

- [ ] Is this code complete?
- [ ] Are tests done/modified?
- [ ] Are docs written for this PR? (if applicable)

## 🚨 Is this a breaking change for API clients?
Yes/No

## :squirrel: Deploy notes
For example: Need to run migrations as part of deployment

## Anything else

Feverfew commented on backend/uclapi/uclapi/urls.py:
```diff
@@ -26,6 +27,8 @@
     url(r'^about', about),
+    url(r'^settings', settings),
+    url(r'logout$', logout),
     url(r'^marketplace', marketplace),
```
Ok, looking at the diff I've noticed I've noticed an issue.

First of all,  all these regexes I've commented on are wrong. Note that in Python regexes have the following format defined [here](https://docs.python.org/3/library/re.html). In particular you'll notice that `^` matches the start of the input and `$` matches the end of the input. Currently you'll be taken to the about page if you navigate to `about`, correctly, but also to `about123`, incorrectly. The inverse happens with `logout`, I will be taken the logout page if I navigate to `i-dont-want-to-logout`, ironically. The solution? All these regexes highlighted above require a `^` at the beginning and a `$` at the end. 

HOWEVER, only correct for the lines changed in this current diff. For the other lines that aren't changed please fix them in a separate PR. I would like to note that I believe `url()` is deprecated since Django 2.0 (which thanks to Zak, we've finally switched to), and `path()` is recommended instead. You can find a way to switch to `path()` [here](https://consideratecode.com/2018/05/02/django-2-0-url-to-path-cheatsheet/). I'll accept a diff that either uses `url()` correctly or switches to `path()` (also correctly), but `path() would be prefferred.

TuxedoFish replied:
Okay have switchedd to using Path for urls in this pull request and I will update the other ones in a future pull request too. Thanks for the heads up!

Feverfew commented on frontend/src/components/layout/NavBar.jsx:
```diff
@@ -18,7 +18,7 @@ import Link from './Link.jsx'
 const links = [
   {
     name: `settings`,
-    link: `/oauth/myapps`,
```
@hueyy @anushks, noting my comment above about the incorrect regex in our Python code, is this doing what we expect from a React PoV?

hueyy replied:
These values go into the href attributes of anchor tags. No regex at work here, so looks fine.

Feverfew commented on backend/uclapi/oauth/urls.py:
```diff
@@ -4,7 +4,6 @@
     url(r'myapps/shibcallback$', views.myapps_shibboleth_callback),
-	url(r'myapps', views.my_apps),
     url(r'authorise/$', views.authorise),
     url(r'shibcallback', views.shibcallback),
```
As a note for a new diff to fix the issue I described earlier, this only requires a `$`, assuming you keep using the `url()` function. If you switch to `path()`, hopefully the document I linked should describe what you should do.

Feverfew commented on backend/uclapi/oauth/urls.py:
```diff
@@ -4,7 +4,6 @@
 
 urlpatterns = [
     url(r'myapps/shibcallback$', views.myapps_shibboleth_callback),
-	url(r'myapps', views.my_apps),
```
Why is this being removed?

TuxedoFish replied:
Moving the settings page from /oauth/myapps => /settings because it makes more sense logically based on our navigation

Feverfew commented on backend/uclapi/oauth/views.py:
```diff
@@ -763,6 +763,10 @@ def deauthorise_app(request):
 
 @ensure_csrf_cookie
 def logout(request):
```
I've been thinking about the point of logout, especially considering issue #970. The conclusion from the issue, AFAICT is that we shouldn't do logout, as it is misleading from a security PoV.

Considering the security aspect, I think we should remove the logout button, view, and url pattern completely. People will logout, and then may give their laptop to someone else, assuming they can't login, when in fact they can. If we remove the option to logout, then giving your laptop to someone else is a security flaw committed by the user willingly, knowing the dangers.

hueyy replied:
Library has the same issue. Their solution is to ask people to close their browser window after they've logged out. This works presumably because shibb cookies are session cookies and expire once the browser session closes.

Perhaps we can keep the logout button but include a warning explaining that users aren't really logged out until they close their browser: e.g.

>You have been logged out of UCL API. However, you are still logged in to other UCL services. To log out completely, please close your web browser.

TuxedoFish replied:
I am happy to do either, I feel like i we add a warning explaining that you need to close your browser it at least gives the user the option to be able to log out and instructions on the website without clearing all of their cookies.

I can add this in later, do you think the warning that they need to close the browser to completely log out is sufficient @Feverfew

TuxedoFish commented:
Okay, I have changed the url => path for the urls that I have changeed in this pull request. I will change the others in a seperate pull request.

I have added a pop up warning when you log out like this:

![image](https://user-images.githubusercontent.com/3999881/76628093-1f890680-6534-11ea-9d18-2a27bf20ce7c.png)

@Feverfew @hueyy do we feel like this solves the security issues

hueyy commented:
I'd prefer if it was a separate page, simply because it's too easy to accidentally close an alert box like this, but this works for now.

TuxedoFish commented:
> I'd prefer if it was a separate page, simply because it's too easy to accidentally close an alert box like this, but this works for now.

Oh okay so kind of like the authorise page and you have to click I understand to go back to the home page?

Feverfew commented on backend/uclapi/uclapi/urls.py:
```diff
@@ -16,8 +16,9 @@
+from oauth.views import settings, logout
 from marketplace.views import marketplace
-
+from django.urls import path
```
Please move this line to just under `from django.contrib import admin`

TuxedoFish replied:
Done!

TuxedoFish commented:
> Ok much better.
> Please address the following, then I'll approve:
> 
> 1. Address my one new inline comment (that isn't resolved).
> 2. Remove the alert (and keep logout): this should be discussed in #970 and implemented in a different PR.
> 
> In addition, with the other URLs that I've noticed are incorrect, please correct them in a separate PR.

Okay perfect, I have disabled the alert and moved the import statement, going to change the urls=>path

Feverfew commented on frontend/src/pages/HomePage.jsx:
```diff
@@ -37,6 +37,12 @@ class HomePage extends React.Component {
+      //alert("You have been logged out of UCL API. However, you are still logged in to "
+      //  + "other UCL services. To log out completely, please close your web browser.")
+    }
+
```
Not necessary for this PR I believe

TuxedoFish commented:
@Feverfew Removed all of those ^^

Feverfew commented on backend/uclapi/dashboard/views.py:
```diff
@@ -199,6 +199,7 @@ def home(request):
     except KeyError:
         logged_in = False
 
+
```
Remove extra line

TuxedoFish commented:
@Feverfew Took out the extra line and did squash and merge
